drv/i2c-devices/src/emc2305.rs

```diff
@@ -0,0 +1,334 @@
+    Pwm3Divide = 0x51,
+    Fan3Cfg1 = 0x52,
+    Fan3Cfg2 = 0x53,
+    Gain3 = 0x55,
+    Fan3SpinUpCfg = 0x56,
+    Fan3MaxStep = 0x57,
+    Fan3MinDrive = 0x58,
+    Fan3ValidTach = 0x59,
+    Fan3DriveFailBandLo = 0x5a,
+    Fan3DriveFailBandHi = 0x5b,
+    Tach3TargetLo = 0x5c,
+    Tach3TargetHi = 0x5d,
+    Tach3ReadingHi = 0x5e,
+    Tach3ReadingLo = 0x5f,
+
+    Fan4Setting = 0x60,
+    Pwm4Divide = 0x61,
+    Fan4Cfg1 = 0x62,
+    Fan4Cfg2 = 0x63,
+    Gain4 = 0x65,
+    Fan4SpinUpCfg = 0x66,
+    Fan4MaxStep = 0x67,
+    Fan4MinDrive = 0x68,
+    Fan4ValidTach = 0x69,
+    Fan4DriveFailBandLo = 0x6a,
+    Fan4DriveFailBandHi = 0x6b,
+    Tach4TargetLo = 0x6c,
+    Tach4TargetHi = 0x6d,
+    Tach4ReadingHi = 0x6e,
+    Tach4ReadingLo = 0x6f,
+
+    Fan5Setting = 0x70,
+    Pwm5Divide = 0x71,
+    Fan5Cfg1 = 0x72,
+    Fan5Cfg2 = 0x73,
+    Gain5 = 0x75,
+    Fan5SpinUpCfg = 0x76,
+    Fan5MaxStep = 0x77,
+    Fan5MinDrive = 0x78,
+    Fan5ValidTach = 0x79,
+    Fan5DriveFailBandLo = 0x7a,
+    Fan5DriveFailBandHi = 0x7b,
+    Tach5TargetLo = 0x7c,
+    Tach5TargetHi = 0x7d,
+    Tach5ReadingHi = 0x7e,
+    Tach5ReadingLo = 0x7f,
+
+    SoftwareLock = 0xef,
+    ProductFeatures = 0xfc,
+    ProductId = 0xfd,
+    MfgId = 0xfe,
+    Revision = 0xff,
+}
+
+pub struct Emc2305 {
+    pub device: I2cDevice,
+}
+
+pub const MAX_FANS: u8 = 5;
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct Fan(u8);
+
+impl TryFrom<u8> for Fan {
+    type Error = ();
+    /// Fans are based on a 0-based index. This should *not* be the number
+    /// of the fan (the fan numbers have a 1-based index)
+    fn try_from(index: u8) -> Result<Self, Self::Error> {
+        if index >= MAX_FANS {
+            Err(())
+        } else {
+            Ok(Self(index))
+        }
+    }
+}
+
+impl Fan {
+    fn register(&self, base: Register) -> Register {
+        Register::from_u8((base as u8) + self.0 * 0x10).unwrap()
+    }
+
+    fn configuration1(&self) -> Register {
+        self.register(Register::Fan1Cfg1)
+    }
+
+    fn configuration2(&self) -> Register {
+        self.register(Register::Fan1Cfg2)
+    }
+
+    fn tach_count(&self) -> Register {
+        self.register(Register::Tach1ReadingHi)
+    }
+
+    fn pwm_target(&self) -> Register {
+        self.register(Register::Fan1Setting)
+    }
+}
+
+fn read_reg8(
+    device: &I2cDevice,
+    register: Register,
+) -> Result<u8, ResponseCode> {
+    device.read_reg::<u8, u8>(register as u8)
+}
+
+fn read_reg16(
+    device: &I2cDevice,
+    register: Register,
+) -> Result<[u8; 2], ResponseCode> {
+    device.read_reg::<u8, [u8; 2]>(register as u8)
+}
+
+fn write_reg8(
+    device: &I2cDevice,
+    register: Register,
+    val: u8,
+) -> Result<(), ResponseCode> {
+    device.write(&[register as u8, val])
+}
+
+fn write_reg16(
+    device: &I2cDevice,
+    register: Register,
+    val: u16,
+) -> Result<(), ResponseCode> {
+    device.write(&[register as u8, (val >> 8) as u8, (val & 0xff) as u8])
```

Comment:
nit, take it or leave it: could also be
```suggestion
    let [lo, hi] = u16::to_le_bytes(val);
    device.write(&[register as u8,  lo, hi])
```

although i have no idea which one generates nicer assembly...